taichi/codegen/codegen_llvm.cpp

```diff
@@ -333,8 +333,13 @@ void CodeGenLLVM::visit(UnaryOpStmt *stmt) {
         from_size = data_type_size(from);
       }
       if (from_size < data_type_size(to)) {
-        llvm_val[stmt] = builder->CreateSExt(
-            llvm_val[stmt->operand], tlctx->get_data_type(stmt->cast_type));
+        if (is_signed(from)) {
+          llvm_val[stmt] = builder->CreateSExt(
+              llvm_val[stmt->operand], tlctx->get_data_type(stmt->cast_type));
+        } else {
+          llvm_val[stmt] = builder->CreateZExt(
+              llvm_val[stmt->operand], tlctx->get_data_type(stmt->cast_type));
+        }
```

Comment:
It seems that the logic may be nastier than this:
```C++
#include <iostream>

int main() {
    using i32 = int;
    using u32 = unsigned;
    using i8 = char;
    using u8 = unsigned char;

    std::cout << i32(i8(-128)) << std::endl;
    std::cout << i32(u8(-128)) << std::endl;
    std::cout << u32(i8(-128)) << std::endl;
    std::cout << u32(u8(-128)) << std::endl;
}
```

Output:
```
-128
128
4294967168
128
```
There are three types of outputs here.

Reply:
Thanks a lot for your very detailed examples. After analyzing them, I guess the logic here might still be :
> use `zext` for unsigned int, and `sext` for signed int.

To valid this, I added these test cases into `test_cast.py`, and the result showed no error. 

In fact, `i32(i8(-128))`(which is -128) and `u32(i8(-128))`(which is 4294967168) are exactly the same in memory(0xFFFFFF80). I think the difference is from how we explain it. If so, is it feasible to extend integers in current way?

Please feel free to point out the mistakes that I made. Thanks you very much! :-)